File: Go(lang)/Coffee Machine/coffeeMachine.go
package main

import "fmt"

type coffeeMachine struct {
	water int
	milk  int
	beans int
	sugar int
	sCups int
	mCups int
	lCups int
	money int
}

func main() {
	machine := &coffeeMachine{
		water: 400,
		milk:  540,
		sugar: 100,
		beans: 120,
		sCups: 9,
		mCups: 9,
		lCups: 9,
		money: 550,
	}

	startAction(machine)
}

func startAction(machine *coffeeMachine) {
	fmt.Println("Write action (buy, fill, take, remaining, exit):")

	for {
		var choice string
		_, err := fmt.Scanf("%s", &choice)
		if err != nil {
			fmt.Println(err)
		}

		switch choice {
		case "buy":
			chooseCoffee(machine)
		case "fill":
			fillMachine(machine)
		case "take":
			takeMoney(machine)
		case "remaining":
			printCurrentIngredients(machine)
		case "exit":
			return
		}
	}
}

func printCurrentIngredients(machine *coffeeMachine) {
	fmt.Println("The coffee machine has:")
	fmt.Printf("%d ml of water\n", machine.water)
	fmt.Printf("%d ml of milk\n", machine.milk)
	fmt.Printf("%d g of coffee beans\n", machine.beans)
	fmt.Printf("%d g of sugar\n", machine.sugar)
	fmt.Printf("%d small disposable cups\n", machine.sCups)
	fmt.Printf("%d middle disposable cups\n", machine.mCups)
	fmt.Printf("%d large disposable cups\n", machine.lCups)
	fmt.Printf("$%d of money\n", machine.money)
}

func makeCoffee(water, milk, beans, sugar, sCups, mCups, lCups, money int, machine *coffeeMachine) {
	isMissing, ingredient := missingIngredient(water, milk, beans, sugar, sCups, mCups, lCups, machine)
	if isMissing {
		fmt.Printf("Sorry, not enough %s!\n", ingredient)
	} else {
		fmt.Println("I have enough resources, making you a coffee!")

		machine.water -= water
		machine.milk -= milk
		machine.beans -= beans
		machine.sugar -= sugar
		machine.sCups -= sCups
		machine.mCups -= mCups
		machine.lCups -= lCups
		machine.money += money
	}
}

func missingIngredient(water, milk, beans, sugar, sCups, mCups, lCups int, machine *coffeeMachine) (bool, string) {
	if machine.water < water {
		return true, "water"
	}
	if machine.milk < milk {
		return true, "milk"
	}
	if machine.beans < beans {
		return true, "beans"
	}
	if machine.sugar < sugar {
		return true, "sugar"
	}
	if machine.sCups < sCups {
		return true, "small cups"
	}
	if machine.mCups < mCups {
		return true, "medium cups"
	}
	if machine.lCups < lCups {
		return true, "large cups"
	}

	return false, ""
}

func fillMachine(machine *coffeeMachine) {
	var tmpVal int

	fmt.Println("Write how many ml of water you want to add:")
	machine.water += getValueForFill(tmpVal)
	fmt.Println("Write how many ml of milk you want to add:")
	machine.milk += getValueForFill(tmpVal)
	fmt.Println("Write how many grams of coffee beans you want to add:")
	machine.beans += getValueForFill(tmpVal)
	fmt.Println("Write how many grams of sugar you want to add:")
	machine.sugar += getValueForFill(tmpVal)
	fmt.Println("Write how many small disposable cups you want to add:")
	machine.sCups += getValueForFill(tmpVal)
	fmt.Println("Write how many medium disposable cups you want to add:")
	machine.mCups += getValueForFill(tmpVal)
	fmt.Println("Write how many large disposable cups you want to add:")
	machine.lCups += getValueForFill(tmpVal)
}

func getValueForFill(fillType int) int {
	_, err := fmt.Scanf("%d", &fillType)
	if err != nil {
		fmt.Println(err)
	}

	return fillType
}

func takeMoney(machine *coffeeMachine) {
	fmt.Printf("I gave you $%d\n", machine.money)
	machine.money = 0
}

func chooseSize() string {
	fmt.Println("Choose your size? 1 - small, 2 - medium, 3 - large, back - to main menu:")

	var size string
	_, err := fmt.Scanf("%s", &size)
	if err != nil {
		fmt.Println(err)
	}

	return size
}

func chooseSugar() string {
	fmt.Println("Do you want sugar? yes, no, back - to main menu:")

	var sugarChoice string
	_, err := fmt.Scanf("%s", &sugarChoice)
	if err != nil {
		fmt.Println(err)
	}

	return sugarChoice
}

func chooseCoffee(machine *coffeeMachine) {
	fmt.Println("What do you want to buy? 1 - espresso, 2 - latte, 3 - cappuccino, back - to main menu:")

	var coffeeTyp string
	_, err := fmt.Scanf("%s", &coffeeTyp)
	if err != nil {
		fmt.Println(err)
	}

	var sugar int
	switch chooseSugar() {
	case "yes":
		sugar = 0
	case "no":
		sugar = 3
	case "back":
		return
	}

	var sc, mc, lc int
	switch chooseSize() {
	case "1":
		sc, mc, lc = 1, 0, 0
	case "2":
		sc, mc, lc = 0, 1, 0
	case "3":
		sc, mc, lc = 0, 0, 1
	case "back":
		return
	}

	switch coffeeTyp {
	case "1":
		makeCoffee(250, 0, 16, sugar, sc, mc, lc, 4, machine)
	case "2":
		makeCoffee(350, 75, 20, sugar, sc, mc, lc, 7, machine)
	case "3":
		makeCoffee(200, 100, 12, sugar, sc, mc, lc, 6, machine)
	case "back":
		return
	}
}

File: Python/Coffee Machine/coffee_machine.py
class CoffeeMachine:
    def __init__(self):
        self.water = 400
        self.milk = 540
        self.coffee_beans = 120
        self.cups = 9
        self.money = 550
        self.status = 'default'

    def handle_states(self, text):
        match self.status:
            case'default':
                self.default_menu(text)
            case 'buy':
                self.buy_menu(text)
                self.status = 'default'
            case 'fill_water':
                self.water += int(text)
                self.status = 'fill_milk'
            case 'fill_milk':
                self.milk += int(text)
                self.status = 'fill_coffee_beans'
            case 'fill_coffee_beans':
                self.coffee_beans += int(text)
                self.status = 'fill_cups'
            case 'fill_cups':
                self.cups += int(text)
                print('')
                self.status = 'default'
            case _:
                print('Wrong option!')

    def default_menu(self, text):
        match text:
            case 'buy':
                self.status = 'buy'
            case 'fill':
                self.status = 'fill_water'
            case 'take':
                self.take_money()
                self.status = 'default'
            case 'remaining':
                self.show_remaining()
                self.status = 'default'
            case 'exit':
                self.status = 'exit'
            case _:
                print('Wrong option!')
        print('')

    def buy_menu(self, text):
        match text:
            case '1':
                self.make_coffee(250, 0, 16, 1, 4)  # espresso
            case '2':
                self.make_coffee(350, 75, 20, 1, 7)  # latte
            case '3':
                self.make_coffee(200, 100, 12, 1, 6)  # cappuccino
            case _:
                print('Wrong option!')
        print('')

    def make_coffee(self, water, milk, coffee_beans, cups, money):
        if self.water < water:
            print('Sorry, not enough water!')
        elif self.milk < milk:
            print('Sorry, not enough milk!')
        elif self.coffee_beans < coffee_beans:
            print('Sorry, not enough coffee beans!')
        elif self.cups < cups:
            print('Sorry, not enough disposable cups!')
        else:
            print('I have enough resources, making you a coffee!')
            self.water -= water
            self.milk -= milk
            self.coffee_beans -= coffee_beans
            self.cups -= cups
            self.money += money

    def fill_machine(self, water, milk, coffee_beans, cups):
        self.water += water
        self.milk += milk
        self.coffee_beans += coffee_beans
        self.cups += cups

    def take_money(self):
        print(f'I gave you ${self.money}')
        self.money = 0
        self.status = 'menu'

    def show_remaining(self):
        print('The coffee machine has:')
        print(f'{self.water} ml of water')
        print(f'{self.milk} ml of milk')
        print(f'{self.coffee_beans} g of coffee beans')
        print(f'{self.cups} disposable cups')
        print(f'${self.money} of money')


def main():
    coffee_machine = CoffeeMachine()

    while coffee_machine.status != 'exit':
        match coffee_machine.status:
            case 'default':
                coffee_machine.handle_states(input('Write action (buy, fill, take, remaining, exit):\n'))
            case 'buy':
                coffee_machine.handle_states(input(
                    'What do you want to buy? 1 - espresso, 2 - latte, 3 - cappuccino, back - to main menu: \n'))
            case 'fill_water':
                coffee_machine.handle_states(input('Write how many ml of water you want to add:\n'))
            case 'fill_milk':
                coffee_machine.handle_states(input('Write how many ml of milk you want to add:\n'))
            case 'fill_coffee_beans':
                coffee_machine.handle_states(input('Write how many grams of coffee beans you want to add:\n'))
            case 'fill_cups':
                coffee_machine.handle_states(input('Write how many disposable cups you want to add:\n'))


if __name__ == "__main__":
    main()
